### ISSUE TITLE
Move modules to the OCA

### ISSUE THREAD

**Issue by robinkeunen** [coopiteasy/vertical-cooperative#359]
Depends on 

- https://github.com/coopiteasy/vertical-cooperative/issues/350
- https://github.com/coopiteasy/vertical-cooperative/issues/227
- https://github.com/coopiteasy/vertical-cooperative/issues/356
- https://github.com/coopiteasy/vertical-cooperative/issues/357

**Comment by ferdymercury** [comment]
Thanks for your wonderful work on moving towards OCA!

One question, is there a (very rough) time estimate when the migration will be complete? Will it target 14.0 and 12.0? Or also 16.0 once it is released?

Thanks again, regards.

**Comment by robinkeunen** [comment]
Hey @ferdymercury ,

- cooperators modules in v14 are here : https://github.com/coopiteasy/vertical-cooperative/tree/14.0
- we're moving modules to the oca/cooperative repository starting this week
- 16.0 is not yet planned, it will depend on the customers we get. If you want to start working on it, please feel free, we will review eagerly :-)

**Comment by ferdymercury** [comment]
> * cooperators modules in v14 are here : https://github.com/coopiteasy/vertical-cooperative/tree/14.0

Thanks for the info! Do you know by chance if the obeesdoo modules are also available for v14?
We are now using all these from odoo v12:

- beesdoo_account
- beesdoo_base
- beesdoo_easy_my_coop
- beesdoo_pos
- beesdoo_pos_coop_status
- beesdoo_product
- beesdoo_shift
- beesdoo_shift_attendance
- beesdoo_website_eater
- beesdoo_website_shift
- beesdoo_worker_status
- easy_my_coop
- easy_my_coop_easy_migration
- easy_my_coop_es
- easy_my_coop_website
- easy_my_coop_website_portal

> * we're moving modules to the oca/cooperative repository starting this week

Nice, thanks!

> * 16.0 is not yet planned, it will depend on the customers we get. If you want to start working on it, please feel free, we will review eagerly :-)

Not sure if I can be of any help with it, I don't know what are the steps involved towards it.

**Comment by robinkeunen** [comment]
Nice to hear that our modules are used elsewhere :-)

We won't port the beesdoo modules to v14, we'll jump directly to v16. You probably noticed we're busy splitting all these modules as well. Tell us if you have trouble upgrading those, we might give a few pointers.

To port on v16, you will need to adapt the code of v14 to the new code that will be out next week. Then you must write migration scripts for the data. We can also arrange a call sometimes to get you started.

**Comment by ferdymercury** [comment]
Our coop supermarket (in Valencia) is planning to migrate from a closed-source ERP to odoo around the end of the year. The script I am writing just ports the data (excel files) to Odoo via its API. My question is, could I just migrate from the Excel files to Odoo v16? Or your modules won't just be available in Odoo16 unless someone has written migration scripts for them?

**Comment by robinkeunen** [comment]
If you're importing data directly to v16, you won't need migration scripts for you data. You will only need the code available in v16. We will probably port cooperator modules in early 2023 and beesdoo modules later in the year, nothing sure though.

**Comment by ferdymercury** [comment]
Thanks for the clarification and congrats for the OCA v14 migration!

Just in case I find some time in the future, how many hours do you estimate it would take me to port the OCA cooperators module from v14 to v16? I have experience with C++ and Python programming, but none with your modules / addons.

**Comment by robinkeunen** [comment]
I'm afraid it's a bit hard to tell because we don't know much about odoo 16 yet. Plus you'd need to get familiar with the odoo framework first. For all the modules, without the migration scripts, I'd say between 40 and 80 hours ?

**Comment by ferdymercury** [comment]
Ok, good to know, thanks for the info. Then it's something I surely can't do, as I am just volunteering a couple of hours here and there during my free time apart from my normal job :) .

**Comment by robinkeunen** [comment]
| module | [12.0][REM] | [14.0][REM] | [12.0][ADD] | [14.0][ADD] | time | comment |
| :----- | :---------: | :---------: | :---------: | :---------: | :--- | :------ |
| `cooperator` | ☑️ | ☑️ | ✅ | ✅ | 0 | |
| `cooperator_api` | | ✖️ | | ✖️ | 0 | not ported in v14 |
| `cooperator_api_logs` |  | ✖️ | | ✖️ | 0 | idem |
| `cooperator_portal` | ☑️ | ☑️ | ✅ | ✅ | 0 | |
| `cooperator_website` | ☑️ | ☑️ | ✅ | ✅ | 0 | |
| `cooperator_website_recaptcha` | | | | | ? | |
| `cooperator_website_uppercase_lastname` | ✖️ | ✖️ | ✖️ | ✖️ | 0 | leave and rename repo to cie-cooperative |
| `easy_my_coop_connector` | ✖️ | ✖️ | ✖️ | ✖️ | 0 | not working |
| `easy_my_coop_dividend` | ✖️ | ✖️ | ✖️ | ✖️ | 0 | not installable |
| `easy_my_coop_export_xlsx` | ✖️ | ✖️ | ✖️ | ✖️ | 0 | not installable |
| `easy_my_coop_loan` | ✖️ | ✖️ | ✖️ | ✖️ | 0 | not ported, move out of cooperator app |
| `easy_my_coop_loan_account` | ✖️ | ✖️ | ✖️ | ✖️ | 0 | idem |
| `easy_my_coop_loan_account_be` | ✖️ | ✖️ | ✖️ | ✖️ | 0 | idem |
| `easy_my_coop_loan_bba` | ✖️ | ✖️ | ✖️ | ✖️ | 0 | idem |
| `easy_my_coop_loan_website` | ✖️ | ✖️ | ✖️ | ✖️ | 0 | idem |
| `l10n_be_cooperator` | ☑️ | ☑️ | ☑️ | ☑️ | 0 | oca/l10n-belgium |
| `l10n_be_cooperator_portal` | | | | | ? | oca/l10n-belgium |
| `l10n_ch_cooperator` | | | | | ? | oca/l10n-switzerland |
| `l10n_es_cooperator` | | | | | ? |oca/l10n-spain |
| `l10n_fr_cooperator` | | | | | ? | oca/l10n-france |
| `portal_recaptcha` | | | ☑️ | | ? | oca/website |

**Legend**

- [12.0][REM]: PR to remove module from coopiteasy/vertical-cooperative on branch 12.0 
- [14.0][REM]: PR to remove module from coopiteasy/vertical-cooperative on branch 14.0 
- [12.0][REM]: PR to add module to oca/cooperative on branch 12.0
- [14.0][REM]: PR to add module to oca/cooperative on branch 14.0

- PR opened : ☑️
- PR merged : ✅
- nothing to do : ✖️

@huguesdk fyi

**Comment by huguesdk** [comment]
moving some modules to other repos might need some extra work. for example, OCA/website#914 is currently blocked because it lowers test coverage, so it has been flagged as “needs tests”.

**Comment by huguesdk** [comment]
## current status

*   OCA/l10n-belgium#150 is blocked by failing tests due to the fact that `cooperator` removes the unique constraint on bank accounts. coopiteasy/vertical-cooperative#310 must be re-created in [OCA/cooperative](https://github.com/OCA/cooperative) (in 12.0 and 14.0) and a migration script must be added to re-enable the unique constraint.
*   OCA/website#914 is blocked by missing tests (it lowers coverage). a test for [this function](https://github.com/coopiteasy/website/blob/12.0-add-portal_recaptcha/portal_recaptcha/models/portal_mixin.py#L27) must be added.
*   `cooperator_website_recaptcha` is blocked by OCA/website#914.
*   `l10n_be_cooperator_portal` is blocked by OCA/l10n-belgium#150.

here’s the order in which i think things should be done:

1.  fix the unique constraint on bank accounts in `cooperator` (coopiteasy/vertical-cooperative#310)
2.  check that tests on OCA/l10n-belgium#150 now pass and add the same pull request for 14.0.
3.  add a test for OCA/website#914 to ensure test coverage stays the same, then add the same pull request for 14.0.
4.  create the pull requests for `cooperator_website_recaptcha` and `l10n_be_cooperator_portal`.

regarding the other modules (api & other l10n), i think that the pull requests should be able to be created at any time.

**Comment by huguesdk** [comment]
## current status

everything stated in the previous message is resolved. `portal_recaptcha` has been given tests and a refactoring, which also needed a refactoring in `cooperator_website_recaptcha`. here are the pending pull requests:

*   OCA/website#914 #407
*   OCA/website#928 #408
*   OCA/cooperative#43 #409
*   OCA/cooperative#44 #410
*   OCA/cooperative#58 OCA/l10n-belgium#181
*   OCA/cooperative#59 OCA/l10n-belgium#182
*   OCA/cooperative#60 #411
*   OCA/cooperative#61 #412

**Comment by robinkeunen** [comment]
@huguesdk thx for the summary !

**Comment by DaniilDigtyar** [comment]
Opened PR to move `l10n_es_cooperator` to `l10n-spain`

- https://github.com/OCA/l10n-spain/pull/2810

**Comment by DaniilDigtyar** [comment]
PR was rejected by `l10n-spain` and say that we move it to `OCA/cooperator` new PR was opened

- https://github.com/OCA/cooperative/pull/51

**Comment by ferdymercury** [comment]
> Nice to hear that our modules are used elsewhere :-)
> 
> We won't port the beesdoo modules to v14, we'll jump directly to v16. You probably noticed we're busy splitting all these modules as well. Tell us if you have trouble upgrading those, we might give a few pointers.

Hi @robinkeunen, sorry to bother again. We are finally starting the ERP system using odoo12 (with the aforementioned beescoop and coopiteasy modules), however I noticed that our Toledo-weighing-scale seems only to be supported in odoo13 (https://github.com/odoo/odoo/blob/13.0/addons/hw_drivers/drivers/SerialScaleDriver.py). I was wondering whether any other of your customers was able to backport that driver into odoo12, or if you have an estimate on how many hours such an endeavour would take. Thanks in advance and congrats for the great advances with the splitting and OCA porting!

**Comment by robinkeunen** [comment]
If not too late, I would strongly recommend going on more recent versions of odoo (16 or at least 14). Migration later will be very costly. We're starting the analysis to move our supermarket towards version 16.

@huguesdk can you update ferdymercury on "the state of the Toledo" ? I'm not sure where we're at right now.

**Comment by ferdymercury** [comment]
> If not too late, I would strongly recommend going on more recent versions of odoo (16 or at least 14). Migration later will be very costly. We're starting the analysis to move our supermarket towards version 16.

We are using the beesdoo modules, are these already ready for 14 or 16?

**Comment by huguesdk** [comment]
@ferdymercury the weighing scale driver that you’re referring to are the drivers that run on the iot box (or pos computer with posboxless), not on the odoo server, and using different versions between those is not a problem, as the communication protocol hasn’t changed since version 8.0.

i didn’t know that the code was still available in the 13.0 branch! it was present in version 12.0 before [this commit](https://github.com/odoo/odoo/commit/a6f44ee93e44b9784e6d0333d0284e3b28a35d9e), which removed it.

this said, i do not recommend using the odoo weighing scale driver. it has a terrible latency (sometimes up to several seconds). instead, i recommend to use [pywebdriver](https://github.com/akretion/pywebdriver) (instead of posboxless), which handles the same mettler toledo protocol (8217) as the odoo driver, but with a much lower latency.

**Comment by ferdymercury** [comment]
Thanks! And how would the pywebdriver communicate with odoo12 on the pos computer with the Rs232 connection to the scale? Does it need an extra plugin installed?

**Comment by ferdymercury** [comment]
Thanks! I will look at it

**Comment by huguesdk** [comment]
the setup is quite simple once you understand how it works:

*   the weighing scale is connected to the pos computer with an rs-232 (or usb) connection.
*   pywebdriver is installed on the pos computer and communicates with the weighing scale.
*   a web browser is used on the pos computer to access odoo.
*   the pos in odoo is configured to communicate to the “iot box” on http://localhost:8069.
*   the javascript odoo pos application runs in the browser and communicates with pywebdriver through http on the pos computer itself.

**Comment by huguesdk** [comment]
@robinkeunen i updated [my previous message above](https://github.com/coopiteasy/vertical-cooperative/issues/359#issuecomment-1344492471) with the new pull requests to move back `l10n_be_cooperator*` modules to cooperative instead of l10n-belgium.

**Comment by ferdymercury** [comment]
> * pywebdriver is installed on the pos computer and communicates with the weighing scale

I am having troubles with this step. I can communicate on Windows via COM1 and a serial terminal with the scale, so I am sure the cable works, but I do not see the scale when I open pywebdriver on the web browser on pos status tab. I only see three columns: display (available COM1, COM2,... In red), escpos and windowsPrinters. I tried modifying the config.ini file to add the scale_driver as in the Github example, but to no avail. Are there any guidelines or instructions I can follow? Or a logfile where I can see where the error comes from? Thanks again!
(Btw if i click on the peripheral USB tab, I get a server error)

**Comment by huguesdk** [comment]
are you using pywebdriver 3.0.15 or more recent? because there was a fix for the scale driver for windows in that version. in your config file, in the `[application]` section, `drivers` should contain `scale_driver`, so at least:
```
[application]
drivers = odoo8,scale_driver
```
and the `[scale_driver]` section should be like this:
```
[scale_driver]
protocol_name=mettler_toledo_8217
port=COM1
baudrate=9600
```
i’m not sure how well it works on windows, i never tried it.

**Comment by ferdymercury** [comment]
Thanks for the prompt reply.
Yes, I was using 3.0.18 https://github.com/akretion/pywebdriver/releases/download/v3.0.18/pywebdriver_win64_installer.exe

and had added exactly those lines to the config file and had re-run install.bat to restart the service. 

Is a fourth column expected to appear in the web browser "status tab" showing the picture of a scale? I was just seeing the three aforementioned columns.

**Comment by huguesdk** [comment]
in the status tab, there should be a `div` for each peripheral, so yes, there should be one for the scale. try using only the scale driver maybe? you’re mentioning a “display” driver using also a serial connection. maybe it is in conflict with the scale driver?

anyway, this github issue is not meant for pywebdriver support. if you think that there is a bug, please file an issue on [akretion/pywebdriver](https://github.com/akretion/pywebdriver/issues). we can also possibly offer paid support if you would like us to look into the issue.
